src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -1,67 +1,211 @@
       .build()
   }
 
+  private def getVmServiceAccountCredential: Credential = {
+    new GoogleCredential.Builder()
+      .setTransport(httpTransport)
+      .setJsonFactory(jsonFactory)
+      .setServiceAccountId(dc.serviceAccount)
+      .setServiceAccountScopes(vmScopes.asJava)
+      .setServiceAccountPrivateKeyFromPemFile(serviceAccountPemFile)
+      .build()
+  }
 
   def createCluster(googleProject: String, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
-    val op = build(googleProject, clusterName, clusterRequest)
-    op.map{op =>
-      val metadata = op.getMetadata
-      ClusterResponse(clusterName, googleProject, metadata.get("clusterUuid").toString, metadata.get("status").toString, metadata.get("description").toString, op.getName)}
+    val bucketName = s"${clusterName}-${UUID.randomUUID.toString}"
+
+    val bucketResponse = initializeBucket(googleProject, clusterName, bucketName, clusterRequest.serviceAccount)
+
+    bucketResponse.flatMap[ClusterResponse]{ Bucket =>
+      val op = build(googleProject, clusterName, clusterRequest, bucketName)
+      op.map { op =>
+        val metadata = op.getMetadata
+        ClusterResponse(clusterName, googleProject, metadata.get("clusterUuid").toString, metadata.get("status").toString, metadata.get("description").toString, op.getName)
+      }
+    }
+  }
+
+  private def createBucketUri(bucketName: String, fileName: String): String = {
+    s"gs://${bucketName}/${fileName}"
+  }
+
+  private def initializeBucket(googleProject: String, clusterName: String, bucketName:String, serviceAccount: String): Future[Bucket] = {
+    import org.broadinstitute.dsde.workbench.leonardo.model.LeonardoJsonSupport._
+    import spray.json._
+    Future {
+      val bucket = new Bucket().setName(bucketName)
+      val bucketInserter = getStorage(getBucketServiceAccountCredential).buckets().insert(googleProject, bucket)
+      try {
+       executeGoogleRequest(bucketInserter) //returns a Bucket
+      } catch {
+        case e: GoogleJsonResponseException => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Google Request Failed: ${e.getDetails.getMessage}"))
+      } finally {
+        val initScriptRaw = scala.io.Source.fromFile(dc.jupyterConfigFolderPath + dc.initActionsScriptName).mkString
+        val replacements =  ClusterInitValues(clusterName, googleProject, dc.dataprocDockerImage,
+                  dc.jupyterProxyDockerImage, createBucketUri(bucketName, dc.jupyterServerCrtName), createBucketUri(bucketName, dc.jupyterServerKeyName),
+                  createBucketUri(bucketName, dc.jupyterRootCaPemName), createBucketUri(bucketName, dc.clusterDockerComposeName), dc.jupyterServerName, dc.proxyServerName).toJson.asJsObject.fields
+        val initScript = replacements.foldLeft(initScriptRaw)((a, b) => a.replaceAllLiterally("$(" + b._1 +")", b._2.toString()))
```

Comment:
Works for me! A comment might make this more clear though. Something like "replace templated values in the raw init script with actual values before populating the bucket."

Reply:
i think a generified version of this templating stuff is a good candidate for workbench-libs